modules/documents/app/services/documents/oauth/generate_token_service.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Documents
+  module OAuth
+    class GenerateTokenService < BaseServices::BaseCallable
+      def initialize(user:)
+        super()
+
+        @user = user
+      end
+
+      def perform
+        application_result = EnsureApplicationService.new.call
+        return application_result unless application_result.success?
+
+        application = application_result.result
+
+        token = create_access_token(application)
+
+        if token.persisted?
+          ServiceResult.success(result: token)
+        else
+          ServiceResult.failure(errors: token.errors)
+        end
+      end
+
+      private
+
+      def create_access_token(application)
+        application.access_tokens.create(
+          resource_owner_id: @user.id,
+          scopes: "api_v3",
+          expires_in: 24.hours.to_i
```

Comment:
I'd question whether such a long expiration time is necessary. The default we configure is two hours.

Without having read the rest of the code: Our implementation should also work well if the expiration time was one minute. Or in other words: The token expiring in the middle of the session should be a case that we consider and that we can deal with.

I am pointing this out, because 24 hours looks like a value chosen to ensure that tokens never expire while we use them.

```suggestion
```

Reply:
Yeah, I agree that this is the way to go, refreshing tokens and all, I just didn't want to do everything in one go. I'd like to merge with 24h, though, and tackle this separately.

I am not sure how to measure what's the real risk of going live with this solution, though, in case we don't implement a refresh flow in time.

Reply:
Let's just make sure to have at least a work package reminding us to do this.

I am not too panicked about the 24 hours, because also our default of two hours is pretty long to be honest. I'd prefer us to bring this down to something in the realm of "a few minutes" at some point.